include/deal.II/base/work_stream.h

```diff
@@ -1229,136 +1378,36 @@ namespace WorkStream
-                              copy_data           = p->copy_data.get();
-                              p->currently_in_use = true;
-                              break;
-                            }
-                        }
-                      // If no element in the list was found, create one and
-                      // mark it as used.
-                      if (scratch_data == nullptr)
-                        {
-                          Assert(copy_data == nullptr, ExcInternalError());
-                          scratch_and_copy_data_list.emplace_back(
-                            std::make_unique<ScratchData>(sample_scratch_data),
-                            std::make_unique<CopyData>(sample_copy_data),
-                            true);
-                          scratch_data = scratch_and_copy_data_list.back()
-                                           .scratch_data.get();
-                          copy_data =
-                            scratch_and_copy_data_list.back().copy_data.get();
-                        }
-
-                      for (const Iterator &it : chunk)
-                        {
-                          if (have_worker)
-                            worker(it, *scratch_data, *copy_data);
-                          if (have_copier)
-                            copier(*copy_data);
-                        }
-
-                      // Mark objects as free to be used again.
-                      for (typename ScratchAndCopyDataList::iterator p =
-                             scratch_and_copy_data_list.begin();
-                           p != scratch_and_copy_data_list.end();
-                           ++p)
-                        {
-                          if (p->scratch_data.get() == scratch_data)
-                            {
-                              Assert(p->currently_in_use == true,
-                                     ExcInternalError());
-                              p->currently_in_use = false;
-                            }
-                        }
-                    })
-                    .name("worker_and_copier");
-                  chunk.clear();
-                  chunk_counter = 0;
-                }
+              tf::Taskflow taskflow;
+
+              tf::IndexRange<int> range(0, color.size(), 1);
+              taskflow.for_each_by_index(
+                range,
+                [&](tf::IndexRange<int> subrange) {
+                  ArrayView<const Iterator> chunk(&color[subrange.begin()],
+                                                  subrange.size());
+                  worker_and_copier(chunk);
+                },
+                tf::GuidedPartitioner(chunk_size));
 
-              if (color > 0)
-                // Wait for the previous color to finish executing
-                execution_future.wait();
-              execution_future = executor.run(std::move(taskflow));
+              executor.run(taskflow).wait();
```

Comment:
I don't know whether it makes a difference, but have you thought first building all tasks in the `taskflow` object, and only run the `executor.run()` function outside the loop over the colors? You'd have to make sure to say that these tasks depend on each other (so that they are not executed in parallel), but at least you'd have to only create one `TaskFlow` object for the whole operation, rather than one per color.

If you have thought about it and have good reasons to leave it as is, that's ok -- I'm just curious :-)